File: src/parser.cpp
#include "lune/parser.hpp"


namespace lune {

namespace {
template <typename T, typename... Args>
std::shared_ptr<T> make_node(Args&&... args) {
    return std::make_shared<T>(T{std::forward<Args>(args)...});
}

ExprPtr make_expr(Expr::Variant node) { return std::make_shared<Expr>(Expr{std::move(node)}); }
StmtPtr make_stmt(Stmt::Variant node) { return std::make_shared<Stmt>(Stmt{std::move(node)}); }
}

Parser::Parser(std::vector<Token> tokens) : tokens_(std::move(tokens)) {}

Program Parser::parse_program() {
    diagnostics_.clear();
    Program program;
    while (!is_at_end()) {
        program.items.push_back(declaration());
    }
    return program;
}

const std::vector<Diagnostic>& Parser::diagnostics() const { return diagnostics_; }

StmtPtr Parser::declaration() {
    if (match(TokenType::KwFn)) return function_declaration();
    if (match(TokenType::KwConst)) return const_declaration();
    return statement();
}

StmtPtr Parser::statement() {
    if (match(TokenType::LBrace)) return block_statement();
    if (match(TokenType::KwIf)) return if_statement();
    if (match(TokenType::KwReturn)) return return_statement();
    return simple_statement();
}

StmtPtr Parser::block_statement() {
    BlockStmt block;
    while (!check(TokenType::RBrace) && !is_at_end()) {
        block.statements.push_back(declaration());
    }
    consume(TokenType::RBrace, "Expected '}' at end of block");
    return make_stmt(block);
}

StmtPtr Parser::if_statement() {
    auto condition = expression();
    consume(TokenType::LBrace, "Expected '{' after if condition");
    auto then_branch = std::get<BlockStmt>(block_statement()->node);

    std::optional<BlockStmt> else_branch;
    if (match(TokenType::KwElse)) {
        consume(TokenType::LBrace, "Expected '{' after else");
        else_branch = std::get<BlockStmt>(block_statement()->node);
    }

    return make_stmt(IfStmt{.condition = condition, .then_branch = std::move(then_branch), .else_branch = std::move(else_branch)});
}

StmtPtr Parser::return_statement() {
    if (check(TokenType::RBrace) || check(TokenType::End)) {
        return make_stmt(ReturnStmt{.expr = std::nullopt});
    }
    return make_stmt(ReturnStmt{.expr = expression()});
}

StmtPtr Parser::function_declaration() {
    const auto& name = consume(TokenType::Identifier, "Expected function name").lexeme;
    consume(TokenType::LParen, "Expected '('");

    std::vector<std::string> params;
    if (!check(TokenType::RParen)) {
        do {
            params.push_back(consume(TokenType::Identifier, "Expected parameter name").lexeme);
        } while (match(TokenType::Comma));
    }
    consume(TokenType::RParen, "Expected ')' after parameters");
    consume(TokenType::LBrace, "Expected function body");
    auto body = std::get<BlockStmt>(block_statement()->node);
    return make_stmt(FunctionDecl{.name = name, .params = std::move(params), .body = std::move(body)});
}

StmtPtr Parser::const_declaration() {
    const auto& name = consume(TokenType::Identifier, "Expected const name").lexeme;
    consume(TokenType::Assign, "Expected '=' in const declaration");
    return make_stmt(ConstDeclStmt{.name = name, .expr = expression()});
}

StmtPtr Parser::simple_statement() {
    if (check(TokenType::Identifier) && tokens_[current_ + 1].type == TokenType::ShortDecl) {
        const auto name = advance().lexeme;
        advance();
        return make_stmt(ShortDeclStmt{.name = name, .expr = expression()});
    }
    if (check(TokenType::Identifier) && tokens_[current_ + 1].type == TokenType::Assign) {
        const auto name = advance().lexeme;
        advance();
        return make_stmt(AssignStmt{.name = name, .expr = expression()});
    }
    return make_stmt(ExprStmt{.expr = expression()});
}

ExprPtr Parser::expression() { return equality(); }

ExprPtr Parser::equality() {
    auto expr = comparison();
    while (match(TokenType::Eq) || match(TokenType::Ne)) {
        const auto op = previous().lexeme;
        expr = make_expr(BinaryExpr{.op = op, .lhs = expr, .rhs = comparison()});
    }
    return expr;
}

ExprPtr Parser::comparison() {
    auto expr = term();
    while (match(TokenType::Lt) || match(TokenType::Le) || match(TokenType::Gt) || match(TokenType::Ge)) {
        const auto op = previous().lexeme;
        expr = make_expr(BinaryExpr{.op = op, .lhs = expr, .rhs = term()});
    }
    return expr;
}

ExprPtr Parser::term() {
    auto expr = factor();
    while (match(TokenType::Plus) || match(TokenType::Minus)) {
        const auto op = previous().lexeme;
        expr = make_expr(BinaryExpr{.op = op, .lhs = expr, .rhs = factor()});
    }
    return expr;
}

ExprPtr Parser::factor() {
    auto expr = unary();
    while (match(TokenType::Star) || match(TokenType::Slash) || match(TokenType::Percent)) {
        const auto op = previous().lexeme;
        expr = make_expr(BinaryExpr{.op = op, .lhs = expr, .rhs = unary()});
    }
    return expr;
}

ExprPtr Parser::unary() {
    if (match(TokenType::Minus)) {
        return make_expr(BinaryExpr{.op = "-", .lhs = make_expr(NumberExpr{0.0}), .rhs = unary()});
    }
    return call();
}

ExprPtr Parser::call() {
    auto expr = primary();
    while (match(TokenType::LParen)) {
        if (!std::holds_alternative<IdentifierExpr>(expr->node)) {
            add_error_here("Only named calls are currently supported");
            return expr;
        }
        CallExpr call_expr{.callee = std::get<IdentifierExpr>(expr->node).name};
        if (!check(TokenType::RParen)) {
            do {
                call_expr.args.push_back(expression());
            } while (match(TokenType::Comma));
        }
        consume(TokenType::RParen, "Expected ')' after call args");
        expr = make_expr(std::move(call_expr));
    }
    return expr;
}

ExprPtr Parser::primary() {
    if (match(TokenType::Number)) return make_expr(NumberExpr{std::stod(previous().lexeme)});
    if (match(TokenType::String)) return make_expr(StringExpr{previous().lexeme});
    if (match(TokenType::KwTrue)) return make_expr(BoolExpr{true});
    if (match(TokenType::KwFalse)) return make_expr(BoolExpr{false});
    if (match(TokenType::KwNull)) return make_expr(NullExpr{});
    if (match(TokenType::Identifier)) return make_expr(IdentifierExpr{previous().lexeme});
    if (match(TokenType::LParen)) {
        auto expr = expression();
        consume(TokenType::RParen, "Expected ')' after expression");
        return expr;
    }
    add_error_here("Expected expression");
    return make_expr(NullExpr{});
}


bool Parser::is_at_end() const { return peek().type == TokenType::End; }
const Token& Parser::peek() const { return tokens_[current_]; }
const Token& Parser::previous() const { return tokens_[current_ - 1]; }
const Token& Parser::advance() {
    if (!is_at_end()) ++current_;
    return previous();
}
bool Parser::check(TokenType type) const { return !is_at_end() && peek().type == type; }
bool Parser::match(TokenType type) {
    if (!check(type)) return false;
    advance();
    return true;
}
const Token& Parser::consume(TokenType type, const char* error_message) {
    if (check(type)) return advance();
    add_error_here(error_message);
    return peek();
}

void Parser::add_error_here(const char* error_message) {
    diagnostics_.push_back(Diagnostic{.message = error_message, .line = peek().line, .column = peek().column});
}

} // namespace lune

File: src/parser.zig
const std = @import("std");
const ast = @import("ast.zig");
const token_mod = @import("token.zig");
const Token = token_mod.Token;
const TokenType = token_mod.TokenType;

const ParseError = error{
    ExpectedTopLevelItem,
    ExpectedIdentifier,
    ExpectedAssign,
    ExpectedLeftParen,
    ExpectedRightParen,
    ExpectedLeftBrace,
    ExpectedRightBrace,
    ExpectedExpression,
    InvalidBinaryOperator,
};

const ParseResult = ParseError || std.mem.Allocator.Error;

pub const Parser = struct {
    tokens: []const Token,
    index: usize = 0,
    allocator: std.mem.Allocator,

    pub fn init(allocator: std.mem.Allocator, tokens: []const Token) Parser {
        return .{ .allocator = allocator, .tokens = tokens };
    }

    pub fn parseProgram(self: *Parser) ParseResult!ast.Program {
        var items = std.ArrayList(ast.Item).init(self.allocator);
        while (!self.current().is(.eof)) {
            try items.append(try self.parseItem());
        }
        return .{ .items = try items.toOwnedSlice() };
    }

    fn parseItem(self: *Parser) ParseResult!ast.Item {
        if (self.match(.kw_const)) return .{ .const_decl = try self.parseConstDecl() };
        if (self.match(.kw_fn)) return .{ .function_decl = try self.parseFunction() };
        return error.ExpectedTopLevelItem;
    }

    fn parseConstDecl(self: *Parser) ParseResult!ast.ConstDecl {
        const name = try self.consume(.identifier, error.ExpectedIdentifier);
        _ = try self.consume(.assign, error.ExpectedAssign);
        const value = try self.parseExpression();
        return .{ .name = name.lexeme, .value = value };
    }

    fn parseFunction(self: *Parser) ParseResult!ast.FunctionDecl {
        const name = try self.consume(.identifier, error.ExpectedIdentifier);
        _ = try self.consume(.l_paren, error.ExpectedLeftParen);

        var params = std.ArrayList([]const u8).init(self.allocator);
        if (!self.current().is(.r_paren)) {
            while (true) {
                const param = try self.consume(.identifier, error.ExpectedIdentifier);
                try params.append(param.lexeme);
                if (!self.match(.comma)) break;
            }
        }

        _ = try self.consume(.r_paren, error.ExpectedRightParen);
        const body = try self.parseBlock();
        return .{ .name = name.lexeme, .params = try params.toOwnedSlice(), .body = body };
    }

    fn parseBlock(self: *Parser) ParseResult![]ast.Stmt {
        _ = try self.consume(.l_brace, error.ExpectedLeftBrace);
        var statements = std.ArrayList(ast.Stmt).init(self.allocator);
        while (!self.current().is(.r_brace) and !self.current().is(.eof)) {
            try statements.append(try self.parseStatement());
        }
        _ = try self.consume(.r_brace, error.ExpectedRightBrace);
        return try statements.toOwnedSlice();
    }

    fn parseStatement(self: *Parser) ParseResult!ast.Stmt {
        if (self.match(.kw_return)) {
            if (self.current().kind == .r_brace) return .{ .return_stmt = null };
            return .{ .return_stmt = try self.parseExpression() };
        }

        if (self.match(.kw_if)) return .{ .if_stmt = try self.parseIfStmt() };
        if (self.match(.kw_while)) return .{ .while_stmt = try self.parseWhileStmt() };

        if (self.current().kind == .identifier and self.peek().kind == .decl_assign) {
            const name = self.advance().lexeme;
            _ = self.advance();
            return .{ .var_decl = .{ .name = name, .value = try self.parseExpression() } };
        }

        if (self.current().kind == .identifier and self.peek().kind == .assign) {
            const name = self.advance().lexeme;
            _ = self.advance();
            return .{ .assign = .{ .name = name, .value = try self.parseExpression() } };
        }

        return .{ .expr_stmt = try self.parseExpression() };
    }

    fn parseIfStmt(self: *Parser) ParseResult!ast.Stmt.if_stmt {
        const condition = try self.parseExpression();
        const then_block = try self.parseBlock();
        var else_block: []ast.Stmt = &[_]ast.Stmt{};
        if (self.match(.kw_else)) else_block = try self.parseBlock();
        return .{ .condition = condition, .then_block = then_block, .else_block = else_block };
    }

    fn parseWhileStmt(self: *Parser) ParseResult!ast.Stmt.while_stmt {
        const condition = try self.parseExpression();
        const body = try self.parseBlock();
        return .{ .condition = condition, .body = body };
    }

    fn parseExpression(self: *Parser) ParseResult!*ast.Expr {
        return self.parseEquality();
    }

    fn parseEquality(self: *Parser) ParseResult!*ast.Expr {
        var expr = try self.parseComparison();
        while (self.match(.equal_equal) or self.match(.bang_equal)) {
            const op_token = self.previous();
            const right = try self.parseComparison();
            expr = try self.makeBinary(op_token.kind, expr, right);
        }
        return expr;
    }

    fn parseComparison(self: *Parser) ParseResult!*ast.Expr {
        var expr = try self.parseTerm();
        while (self.match(.less) or self.match(.less_equal) or self.match(.greater) or self.match(.greater_equal)) {
            const op_token = self.previous();
            const right = try self.parseTerm();
            expr = try self.makeBinary(op_token.kind, expr, right);
        }
        return expr;
    }

    fn parseTerm(self: *Parser) ParseResult!*ast.Expr {
        var expr = try self.parseFactor();
        while (self.match(.plus) or self.match(.minus)) {
            const op_token = self.previous();
            const right = try self.parseFactor();
            expr = try self.makeBinary(op_token.kind, expr, right);
        }
        return expr;
    }

    fn parseFactor(self: *Parser) ParseResult!*ast.Expr {
        var expr = try self.parsePrimary();
        while (self.match(.star) or self.match(.slash) or self.match(.percent)) {
            const op_token = self.previous();
            const right = try self.parsePrimary();
            expr = try self.makeBinary(op_token.kind, expr, right);
        }
        return expr;
    }

    fn parsePrimary(self: *Parser) ParseResult!*ast.Expr {
        if (self.match(.number)) return self.newExpr(.{ .number = self.previous().number.? });
        if (self.match(.kw_true)) return self.newExpr(.{ .boolean = true });
        if (self.match(.kw_false)) return self.newExpr(.{ .boolean = false });

        if (self.match(.identifier)) {
            const name = self.previous().lexeme;
            if (!self.match(.l_paren)) return self.newExpr(.{ .variable = name });

            var args = std.ArrayList(*ast.Expr).init(self.allocator);
            if (!self.current().is(.r_paren)) {
                while (true) {
                    try args.append(try self.parseExpression());
                    if (!self.match(.comma)) break;
                }
            }
            _ = try self.consume(.r_paren, error.ExpectedRightParen);
            return self.newExpr(.{ .call = .{ .name = name, .args = try args.toOwnedSlice() } });
        }

        if (self.match(.l_paren)) {
            const expr = try self.parseExpression();
            _ = try self.consume(.r_paren, error.ExpectedRightParen);
            return expr;
        }

        return error.ExpectedExpression;
    }

    fn makeBinary(self: *Parser, token_kind: TokenType, left: *ast.Expr, right: *ast.Expr) ParseResult!*ast.Expr {
        const op = switch (token_kind) {
            .plus => ast.BinaryOp.add,
            .minus => ast.BinaryOp.sub,
            .star => ast.BinaryOp.mul,
            .slash => ast.BinaryOp.div,
            .percent => ast.BinaryOp.mod,
            .equal_equal => ast.BinaryOp.eq,
            .bang_equal => ast.BinaryOp.neq,
            .less => ast.BinaryOp.lt,
            .less_equal => ast.BinaryOp.lte,
            .greater => ast.BinaryOp.gt,
            .greater_equal => ast.BinaryOp.gte,
            else => return error.InvalidBinaryOperator,
        };
        return self.newExpr(.{ .binary = .{ .op = op, .left = left, .right = right } });
    }

    fn newExpr(self: *Parser, expression: ast.Expr) ParseResult!*ast.Expr {
        const ptr = try self.allocator.create(ast.Expr);
        ptr.* = expression;
        return ptr;
    }

    fn consume(self: *Parser, kind: TokenType, parse_error: ParseError) ParseResult!Token {
        if (self.current().kind == kind) return self.advance();
        return parse_error;
    }

    fn match(self: *Parser, kind: TokenType) bool {
        if (self.current().kind != kind) return false;
        _ = self.advance();
        return true;
    }

    fn current(self: *const Parser) Token {
        return self.tokens[self.index];
    }

    fn peek(self: *const Parser) Token {
        if (self.index + 1 >= self.tokens.len) return self.tokens[self.index];
        return self.tokens[self.index + 1];
    }

    fn previous(self: *const Parser) Token {
        return self.tokens[self.index - 1];
    }

    fn advance(self: *Parser) Token {
        const token = self.tokens[self.index];
        self.index += 1;
        return token;
    }
};
